### ISSUE TITLE
[ADAP-853] [Feature] Support Dynamic Data Masking 

### ISSUE THREAD

**Issue by soksamnanglim** [dbt-labs/dbt-redshift#589]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-redshift functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

I am opening this issue to track the development of [Dynamic Data Masking (DDM)](https://docs.aws.amazon.com/redshift/latest/dg/t_ddm.html) support for dbt-redshift. DDM allows customers to dynamically mask columns for users and roles. 

The following are features of DDM:
* GRANT PERMISSION, USER CREATION, ROLE CREATION/GRANTING
* MASKING POLICY CREATION
* MASKING POLICY ATTACHING
* MASKING POLICY ALTERATION
* MASKING POLICY DETACHMENT AND DROPPING

We want dbt-redshift to support masking policy creation, attachment, alteration, and detachment and dropping. Role granting is currently unsupported by dbt-redshift and would require a separate issue opened for it. At minimum, dbt-redshift should allow users to create and attach masking policies to tables. Ideally, a mechanism to alter, detach, and drop existing masking policies should also exist.

### Describe alternatives you've considered

This [discourse](https://discourse.getdbt.com/t/how-to-implement-dynamic-data-masking-on-redshift/2043) was opened in Jan 2021 before dynamic data masking was a feature in Redshift. As of Nov 2022, Redshift supports DDM, however, users leveraging dbt are unable to manage DDM in their dbt projects.

The alternative solution is to continue using Redshift database drivers or the Redshift Query Editor to manage DDM.

### Who will this benefit?

This feature will be helpful for users (with DDM permissions) who want to manage data masking policies using dbt.

**Comment by dataders** [comment]
@soksamnanglim thanks for writing this up! We discussed this internally and also believe it would be valuable. I'd check out the write up in https://github.com/dbt-labs/dbt-snowflake/issues/531#issuecomment-1700746672. I'd love to hear which approach makes more sense to you.

> Role granting is currently unsupported by dbt-redshift and would require a separate issue opened for it.
Agreed! #415 is where we're tracking that work. IMO, we should ship that feature, before starting this one because:
1. it's a lower lift,
2. makes sense to come first
3. gives us lead time to think through DDM implementation in concert with Core's idea

Looking at the docs you've linked above, two things stand out

1. my understanding is that Redshift's DDM does not support `ALTER`ing existing tables to add masking. This leads me to believe that the implementation will be via [model contracts](https://docs.getdbt.com/docs/collaborate/govern/model-contracts)
2. this line reinforces my view that dbt should not support data sharing
  > DDM policies can't be used with data sharing

**Comment by soksamnanglim** [comment]
@dataders thanks for getting back so quick! 
wrt whether to prioritize role granting or DDM, we are still internally discussing, however personally I am leaning towards shipping #415 first as not supporting end-to-end DDM on dbt would mean users without established RBAC cannot take advantage of DDM. Likewise, DDM seems to be a big feature so time to sort through DDM implementation would be nice 😅 .

> my understanding is that Redshift's DDM does not support ALTERing existing tables to add masking. This leads me to believe that the implementation will be via [model contracts](https://docs.getdbt.com/docs/collaborate/govern/model-contracts)

wrt above, I may be misunderstanding so please correct me! Redshift's DDM can be applied to existing tables. Are you instead referring to this [dbt-labs/dbt-snowflake#531](https://github.com/dbt-labs/dbt-snowflake/issues/531#issuecomment-1700746672)? afaik, dbt-redshift would also leave a window where data is unmasked between table creation and DDM application. Forgive my ignorance—how big of a problem is this? 

I think we can implement DDM without model contracts, unless there is case I'm overlooking? 

> this line reinforces my view that dbt should not support data sharing

We understand your view in the context of DDM. While we are still having an internal discussion, we might also want to consider that DDM is used by users who don't use data sharing and vice versa.

**Comment by soksamnanglim** [comment]
Hi, chiming in again to provide an update that we will be working on allowing dbt users to configure role grants first ([#415](https://github.com/dbt-labs/dbt-redshift/issues/415)). We will keep you updated on DDM!
